scrape/scrape.go

```diff
@@ -1339,13 +1353,22 @@ loop:
 			var ref uint64
 			ref, err = app.Add(lset, t, v)
 			sampleAdded, err = sl.checkAddError(nil, met, tp, err, &sampleLimitErr, &appErrs)
+
 			if err != nil {
 				if err != storage.ErrNotFound {
 					level.Debug(sl.l).Log("msg", "Unexpected error", "series", string(met), "err", err)
 				}
 				break loop
 			}
 
+			if hasExemplar := p.Exemplar(&e); hasExemplar {
+				if err := app.AddExemplar(lset, e); err != nil {
+					if err != storage.ErrDuplicateExemplar {
+						level.Debug(sl.l).Log("msg", "Unexpected error", "error", err, "seriesLabels", lset, "exemplar", e)
+					}
+				}
+			}
+
```

Comment:
Error here is being swallowed again...

I think this API makes it really hard to correctly handle the errors - what if the samples succeeds but the exemplar doesn't?  You probably don't want to hear this at this stage, but I think the `Add` and `AddFast` methods should take exemplars...

Reply:
At this stage I would go ahead with AddExemplar and AddExemplarFast. What you suggested could be a future work based on requirements.

Reply:
Chatted with Callum, and yeah I think this makes sense now.  We have the `Commit` and `Rollback` logic for partial failure.